core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -957,6 +961,12 @@ private[log] class Cleaner(val id: Int,
       if(position == startPosition)
         growBuffersOrFail(segment.log, position, maxLogMessageSize, records)
     }
+
+    // check for missing offsets at the end of logSegment
+    if (lastOffsetInSegment < nextSegmentStartOffset - 1L) {
```

Comment:
@junrao 

well you were right, and after i wrote test case(es), first fix didn't resolve issue, so I've reverted the code, and give 2nd chance to fix it.